### ISSUE TITLE
Request: Allow user-defined storage devices for bookmarking

### ISSUE THREAD

**Issue by daattali** [rstudio/shiny#2537]
When using server-side bookmarking, the state of the app gets saved as an RDS file to the file system. There are cases where this would be almost a perfect solution, except you can't use the local file system for some reason (for example, when you're hosting on infrastructure that doesn't give you persistent local storage). For those cases, it would be great if you could specify your own place to save the RDS file to/load from.

From a very quick glance at the source code, I found that setting `shinyOptions(save.interface = mysave)` and `shinyOptions(load.interface = myload)` allows me to perform bookmarking from a custom directory. But I can only make it work with a directory on the local machine. It looks like it might be feasible to modify the bookmarking source code a little bit to also allow more flexibility to allow the user to save the file to a completely different medium as well, such as S3.

**Comment by ismirsehregal** [comment]
Just if anyone is interested, [here](https://community.rstudio.com/t/changing-bookmark-state-directory-when-saving-to-disk/37877/2) the `shinyOptions(save.interface = ...)` workaround mentioned above is described.

**Comment by pat-alt** [comment]
Just revisiting this as I'm still struggling to restore from bookmarks in a remote location (I originally opened the [thread](https://community.rstudio.com/t/changing-bookmark-state-directory-when-saving-to-disk/37877/2) highlighted by @ismirsehregal above). 

Following @cpsievert 's advice I managed to set `save.interface` such as to save bookmarks in a location specified by the user (`myBookmarks$path`). I set the `save.interface` option just before running `session$doBookmark()`:

```
my_save_interface <- function(id, callback) {
      stateDir <- file.path(myBookmarks$path, "shiny_bookmarks", id)
      if (!shiny:::dirExists(stateDir)) {
        dir.create(stateDir, recursive = TRUE)
      }  
      callback(stateDir)
    }
shinyOptions(save.interface = my_save_interface)
session$doBookmark()
```

I'm struggling to correctly define `load.interface` as on/before restore `myBookmarks$path` is not yet available so I'm not sure how to tell Shiny where to look for the bookmarks. I tried saving the specified path as a `state$value` and defining the `load.interface` option at the same time, i.e. before `session$doBookmark`, so

```
my_save_interface <- function(id, callback) {
      stateDir <- file.path(myBookmarks$path, "shiny_bookmarks", id)
      if (!shiny:::dirExists(stateDir)) {
        dir.create(stateDir, recursive = TRUE)
      }  
      callback(stateDir)
    }
my_load_interface <- function(id, callback) {
      stateDir <- file.path(state$values$path, "shiny_bookmarks", id)
      callback(stateDir)
    }
shinyOptions(save.interface = my_save_interface,
                      load.interface = my_load_interface)
session$doBookmark()
```

but then when trying to restore (perhaps unsurprisingly) object `state` is not available (because no `state` has been loaded yet.

For background, restore is triggered as users click on a hyperlink (e.g. to `"http://127.0.0.1:2020/?_state_id_=af7e123ab925f6de"`) in a url table containing bookmarks and saved `onBookmarked` (similar to [here](https://stackoverflow.com/questions/53855748/save-r-shiny-bookmark-into-a-table-and-retrieve-it)). 

Ideally I'd like to be able specify `load.interface` after the user clicks on the URL and before `RestoreContext` is triggered, so following the same logic as for saving bookmarks:

```
# Stylized code example ----
my_load_interface <- function(id, callback) {
      stateDir <- file.path(myBookmarks$path, "shiny_bookmarks", id)
      if (!shiny:::dirExists(stateDir)) {
        dir.create(stateDir, recursive = TRUE)
      }  
      callback(stateDir)
    }
shinyOptions(save.interface = my_save_interface)
session$doRestore() # imaginary function
```

The underlying issue here is probably just my lack of understanding of what exactly is going on under the bonnet, when a user click on the hyperlink. Realize this is a long comment and may be hard to follow, but hope it makes sense. Any pointers would be much appreciated. 

Many thanks!

Pat

**Comment by daattali** [comment]
Thanks @pat-alt there are certainly ways to work around this limitation of bookmarking. I wouldn't promote this too much though because I think this would be an extremely useful feature for power users and should be supported natively in shiny!

**Comment by see24** [comment]
I have a related question for an app that is running locally. In my case the app is opened each time it is used by running `shinyApp` and then I use the app to do an assessment, save a csv with the results and exit the app. If I come back another day I rerun the app and calculate the value from scratch. What would be really useful is if I could include a save dialogue that would save the state of the app to the local disk, similar to what bookmarking does, and then I could close the app, shut down my computer and come back to the analysis some time later and use another dialogue box to load the state of the app back to where it was when I left.  This does not currently work with bookmarking, I think because restore step only happens when the url is clicked on. Creating the dialogue boxes is easy but would it be possible to have a set of functions ie `bookmarkTo` which would take a location and other bookmark options and save the state to an rds files and `restoreFrom` which would take a location of the rds file and then update the inputs based on the file?  

I have tried using `shiny:::serializeReactiveValues` to save the values in a list but then I would have to use the various updateInput functions to load the state and I am reluctant to do that when it seems like the bookmarking functionality is already set up for that. 

Thanks to @pat-alt for the idea of saving the urls! I will try that if this is a non-starter.

**Comment by ismirsehregal** [comment]
Using `enableBookmarking(store = "server")` already creates local bookmarks (.rds files)

Please check my related answers [here](https://stackoverflow.com/questions/54426953/bookmarking-and-saving-the-bookmarks-in-r-shiny/54435930#54435930) or [here](https://stackoverflow.com/a/68253464/9841389) .

**Comment by see24** [comment]
@ismirsehregal thanks! I had not seen that second answer you linked and I think it will work for my situation. 
However, I do think the feature request still stands. Your answer is a helpful workaround but for a run of the mill user like myself it is quite challenging to understand and use. Since saving is a pretty useful feature in many contexts it would be great if shiny had a more flexible and well documented method.

**Comment by JerePlum99** [comment]
@cpsievert Any idea if this is on the roadmap? Or does anyone know if there is a more modern solution? 

I'm guessing the solution would be similar to the cloud support provided in https://github.com/r-lib/memoise/ or a cache object like in `bindCache`?
